plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -249,28 +271,83 @@ private static boolean isLocationPresent(Storage storage)
         return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
     }
 
-    /**
-     * The class enforces intentionally object identity semantics for the value holder,
-     * not value-based class semantics to correctly act as an invalidation guard in the
-     * cache.
-     */
     private static class ValueHolder
     {
-        private final Optional<List<LocatedFileStatus>> files;
+        private final List<LocatedFileStatus> files = synchronizedList(new ArrayList<>());
+        @GuardedBy("this")
+        @Nullable
+        private RemoteIterator<LocatedFileStatus> fileIterator;
+        @GuardedBy("this")
+        @Nullable
+        private IOException exception;
+
+        public ValueHolder(RemoteIterator<LocatedFileStatus> fileIterator)
+        {
+            this.fileIterator = requireNonNull(fileIterator, "fileIterator is null");
+        }
 
-        public ValueHolder()
+        public synchronized boolean isFullyCached()
         {
-            files = Optional.empty();
+            return fileIterator == null && exception == null;
         }
 
-        public ValueHolder(List<LocatedFileStatus> files)
+        public int getFilesSize()
         {
-            this.files = Optional.of(ImmutableList.copyOf(requireNonNull(files, "files is null")));
+            return files.size();
         }
 
-        public Optional<List<LocatedFileStatus>> getFiles()
+        public List<LocatedFileStatus> getFiles()
         {
-            return files;
+            checkState(isFullyCached());
+            return ImmutableList.copyOf(files);
+        }
+
+        public boolean hasNext(int index)
+                throws IOException
+        {
+            int filesSize = files.size();
+            checkArgument(index >= 0 && index <= filesSize);
+
+            if (index < filesSize) {
+                return true;
+            }
+
+            return fetchNextFile(index);
+        }
+
+        public LocatedFileStatus getFile(int index)
```

Comment:
you're inventing your own API, so it doesn't need to follow Iterator's interface. it would be more efficient (especially with regards to synchronizaiton) to have one method 
```
public synchronized Optional<LocatedFileStatus> getNext(int index);
```